contributors/design-proposals/service_links.md

```diff
@@ -0,0 +1,237 @@
+
+Cons:
+
+1.  Most difficult to explain and reason about
+2.  Most difficult to support
+3.  Higher implementation cost
+4.  Large conceptual delta
+5.  Likely to add significant complexity to pod spec
+6.  Likely to be very verbose
+
+#### Lessons Learned from Prior APIs
+
+One fundamental lesson of Kubernetes API design is that we must make backward
+compatible API changes except in cases of special exception.  This means that we
+need to consider backward compatibility when making new API changes.  As a
+requirement, backward compatibility imposes a burden that you have to live with
+the APIs you release.  This means that we should attempt to be conservative with
+new API features and avoid unnecessary complexity.  Therefore, a simpler API is
+advantageous because complexity can be added later **if desired and warranted**.
+
+#### Middleground
+
+A conservative middleground might be an API that has behavior extremely close to
+the current implicit behavior but allows specifying custom variable names.  This
+approach has the advantage of being close to what users already expect, and can
+be implemented with a fairly simple API.
+
+### Non-existent Services
+
+When a service referenced using this API does not exist, the Kubelet should
+avoid starting the pod.
+
+## Proposed Changes
+
+### API changes
+
+Pull request [#148](https://github.com/kubernetes/community/pull/148)
+proposes an `EnvFrom` addition to the PodSpec to handle creating multiple
+environment variables from a single subject.  This API surface is ideal for the
+service use-cases, so we will extend it here:
+
+```go
+type Container struct {
+	// other fields omitted
+
+	EnvFrom []EnvFromSource `json:"envFrom,omitempty"`
+}
+
+type EnvVarFromSource struct {
+	// other fields omitted
+
+	// Prefix is an optional prefix for generated environment variables
+	Prefix string `json:"prefix,omitempty"`
+
+	// Can be refactored to ObjectReference at a later time while
+	// maintaining backward compatibility.
+	Service *LocalObjectReference `json:"service,omitempty"`
+}
+```
+
+When a pod uses this API feature, it indicates that the automatic injection
+behavior is not desired for that pod.  No automatic injection will occur when
+any container in a pod uses an explicit link to a service.
+
+### Kubelet Changes
+
+The Kubelet must be modified to:
+
+1. Support creation of environment variables via the new API
+2. Disable the automatic injection for a pod whenever an explicit service link
+   appears in the pod's spec
+3. Start the containers in a pod only when all of the pod's explicitly linked
+   services exist
+4. Create an event for the pod when pod start is delayed due to waiting for a
+   service to exist
+
+## Examples
+
+### No custom variable names
+
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: simple-service-reference-example
+spec:
+  containers:
+  - image: busybox
+    name: example-container
+    envFrom:
```

Comment:
@erictune Are you using the service link as a gate for starting? Even if they appear as env vars, it is up to you whether you use them.